core/ledger/snapshotgrpc/snapshot_service.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package snapshotgrpc
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/empty"
+	pb "github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/hyperledger/fabric/core/ledger"
+	"github.com/pkg/errors"
+)
+
+// Snapshot Service implements SnapshotServer grpc interface
+type SnapshotService struct {
+	LedgerGetter LedgerGetter
+}
+
+// LedgerGetter gets the PeerLedger associated with a channel.
+type LedgerGetter interface {
+	GetLedger(cid string) ledger.PeerLedger
+}
+
+// Generate generates a snapshot request.
+func (s *SnapshotService) Generate(ctx context.Context, signedRequest *pb.SignedSnapshotRequest) (*empty.Empty, error) {
+	request := &pb.SnapshotRequest{}
+	if err := proto.Unmarshal(signedRequest.Request, request); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal snapshot request")
+	}
+	if request.ChannelId == "" {
+		return nil, errors.New("missing channel ID")
+	}
+
+	lgr := s.LedgerGetter.GetLedger(request.ChannelId)
+	if lgr == nil {
+		return nil, errors.Errorf("cannot find ledger for channel %s", request.ChannelId)
+	}
+
```

Comment:
as this part of the code is repeating thrice, we can move it into a separate function.

Reply:
The code is not exactly repeated because the proto msg is different for QueryBindings (SnapshotQuery vs. SnapshotRequest). After unmarshalling, there are two parts of common code: channelID validation and getting ledger. We can move them into a common func like the following.

func (s *SnapshotService) getLedger(channelID string) (ledger.PeerLedger, error) {
	if channelID == "" {
		return nil, errors.New("missing channel ID")
	}
	lgr := s.LedgerGetter.GetLedger(channelID)
	if lgr == nil {
		return nil, errors.Errorf("cannot find ledger for channel %s", channelID)
	}
	return lgr, nil
}

Reply:
Yeah, so there is code repetition that can be avoided. That's what I mentioned.